### ISSUE TITLE
Dirichlet distribution sampling not summing to 1

### ISSUE THREAD

**Issue by LekiDuckHasz** [JuliaStats/Distributions.jl#544]
Hello,

Thank you for working on this package.

I'm currently using it to sample Dirichlet distributions. I just remarked that it's not summing to 1 all the time. I fixed it for myself by correcting the last probability being equal to 1 - sum of others.

Sincerely,
LekiDuckHasz

**Comment by ararslan** [comment]
When drawing random values from a distribution, you wouldn't expect the _values_ to sum to 1, you'd expect their respective _probabilities_ to sum to 1. So if you were to generate a lot of random variates from a distribution and evaluate the probability distribution function at each generated value, the sum of the distribution values would be approximately 1. 

Perhaps that's what you're doing and I'm misunderstanding. Can you give a minimum reproducible example of code?

**Comment by LekiDuckHasz** [comment]
Hello,

Thanks for your reply araslan.

I think my request has been misunderstood. I'm specifically talking about Dirichlet and Beta distributions which I use to generate priors for Multinomial and Binomial distributions (resp.).

Try out this code, it should give a sum different from 1 after several times :
`
julia> dir = Dirichlet(ones(5))
julia> x = rand(dir)
julia> sum(x)
`

Sincerely,
LekiDuckHasz

**Comment by johnmyleswhite** [comment]
Could you give an example with a specific number of times and a specific level of violation? As is, this is a really good start at a bug report, but not actually a truly reproducible bug report.

For example, I see things like the following:

```
julia> using Distributions

julia> dir = Dirichlet(ones(5))
Distributions.Dirichlet{Float64}(alpha=[1.0,1.0,1.0,1.0,1.0])

julia> x = rand(dir)
5-element Array{Float64,1}:
 0.0261653
 0.0425999
 0.265175
 0.350671
 0.315389

julia> sum(x)
0.9999999999999999
```

What's an application in which this kind of floating point error is a problem? If we were R, we'd just report that sum as `1.0` and most users would never notice the difference.

**Comment by johnmyleswhite** [comment]
Note that, for example, the following is a property of floating point numbers rather than a property of Julia:

```
julia> 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1
0.9999999999999999
```

**Comment by LekiDuckHasz** [comment]
Hello johnmyleswhite,

Thanks for your reply.

Indeed, I'm quite surprised the following code is not provoking errors : 
```
julia> srand(1)
MersenneTwister(Base.dSFMT.DSFMT_state(Int32[1749029653,1072851681,1610647787,1072862326,1841712345,1073426746,-198061126,1073322060,-156153802,1073567984  …  1977574422,1073209915,278919868,1072835605,1290372147,18858467,1815133874,-1716870370,382,0]),[6.4688e-316,NaN,NaN,1.061e-313,5.02608e180,6.98939e-316,NaN,2.122e-314,4.94066e-323,6.47355e-316  …  NaN,6.36599e-314,6.94062e-310,6.473e-316,NaN,8.48798e-314,6.95293e-310,1.08349e-320,6.94063e-310,6.94063e-310],382,UInt32[0x00000001])

julia> dir = Dirichlet(ones(5))
Distributions.Dirichlet{Float64}(alpha=[1.0,1.0,1.0,1.0,1.0])

julia> x = rand(dir)
5-element Array{Float64,1}:
 0.160101 
 0.0154788
 0.450919 
 0.12886  
 0.244642 

julia> sum(x)
0.9999999999999998

julia> Multinomial(5, x)
Distributions.Multinomial{Float64}(n=5, p=[0.16010097334029538,0.015478839694071832,0.4509187136320659,0.12885994327212083,0.2446415300614459])



```I was expecting to have this kind of result :
```

julia> Multinomial(5, [0.1, 0.2, 0.4, 0.29])
ERROR: ArgumentError: p = [0.1,0.2,0.4,0.29] is not a probability vector.
 in call at /home/bolivier/.julia/v0.4/Distributions/src/multivariate/multinomial.jl:10

```

Though, for the sake of the beauty, I corrected it as follows :

`x[end] = 1 - sum(x[1:end-1])
`

I think this topic can now be closed. Thanks for you replies and remarks guys.

Sincerely,
LekiDuckHasz

**Comment by johnmyleswhite** [comment]
Many people want to be able to sample from non-normalized measures. It's often useful in practice, although there's a good argument against it for error-checking.

I encourage you to read a bit more about floating point numbers so that you can understand why your approach of forcing the summation (via something like `x[end] = 1 - sum(x[1:end-1])`) might produce less accurate results in some cases than the approach we take.

The example I gave above implicitly shows why you should not always do what you're suggesting. To make it explicit:

```
julia> 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1
0.8999999999999999

julia> 1 - sum(0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1)
0.10000000000000009
```

In short, you achieve a sum that superficially seems to be correct at the expense of creating a distribution that's not perfectly uniform.

Another way to see how things change when you force normalization is to work with rational numbers instead of floating point numbers:

```
julia> rationalize(0.1)
1//10

julia> rationalize(1 - sum(0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1))
97375127078282//973751270782819
```

**Comment by brethvoice** [comment]
Does Julia have variable precision options to allow exact rational calculations (at the expense of time and processing power) like the vpi package in MATLAB?

**Comment by simonbyrne** [comment]
@brethvoice There is the `BigInt` type for arbitrary-precision integers. In future though, it would be better to ask this sort of question on [julia-users](https://groups.google.com/forum/#!forum/julia-users).

**Comment by brethvoice** [comment]
@simonbyrne Thanks for letting me know.  I also watched a video tutorial yesterday about arbitrary rational numbers using a double division sign like "//" to ensure the output is also rational rather than a machine calculated number in computer algebra format.  I guess I have a lot to learn about Julia, but I do appreciate your hint about the julia-users Github forum.  I just thought that @johnmyleswhite could solve his problem using variable/exact precision arithmetic such as what you suggested.

**Comment by brethvoice** [comment]
Actually, I guess it was @LekiDuckHasz who had the problem with the rounding error, not @johnmyleswhite.  My mistake in the comment above.
